NARA mapping

markbreedlove opened the pull request:
There are a number of "FIXME" comments in the file to flag things that should be fixed or questions that should be answered before this is merged.

Specifically, I'm not sure that the date handling is correct.

Also:
- The comments have yet to be cleaned up.  I plan to do that before squashing and merging.  I left in the comments that were there when I received the file, which have XML examples from their old datafiles, and these should probably be revised to reflect the JSON of the API.
- I'm not sure about the naming of the `make_*` methods.   I was trying to avoid confusion or naming conflicts with existing methods for the various fields, but I'm not sure that this is the best convention.

anarchivist committed:
NARA WIP

Commit:
Map some easy fields

Map some fields that are the easiest to map.  These tend to be ones that
do not require their own methods for special handling.

Work in progress.  There are FIXME notes that identify issues and
questions.

Commit:
Map some more fields

Commit:
Map relation

Commit:
Fix mapping of description and fix indentation

Commit:
Fill in rights

Commit:
Map dates

Commit:
Avert exception assigning providedLabel

Temporarily fix the return value of `#contributor_term_name` when we
want to say that "nothing" is returned.  See the FIXME comment.  There
is probably a better way to do this.

Commit:
Rearrange comments; fix description field

Commit:
Zero-pad day and month values

Commit:
Add whitespace and parentheses for clarity

Commit:
Fix object and preview

Commit:
Fix object, preview, and data handling of nil results

Commit:
Fix unescaped object URIs; Bandage the date mapping

- Handle object URLs with space characters by escaping them.  Otherwise,
  an exception will be encountered because it's not a valid RDF::URI

- Take out a stray binding.pry statement

- Stop the date mapping from choking on multiple values.  STILL NEEDS
  WORK.

Commit:
fixup - remove binding.pry

Commit:
Fix handling of multiple descriptions.

Commit:
Handle thumbnail URLs with space characters

Commit:
Handle missing identifier type

anarchivist commented:
> I'm not sure about the naming of the make_* methods. I was trying to avoid confusion or naming conflicts with existing methods for the various fields, but I'm not sure that this is the best convention.

I'm wondering if these methods should be namespaced into a module, e.g. something like `Heidrun::MappingHelpers::NARA` (or something more terse?). That could also allow us to separate the mapping file from its helper methods, and could allow us to test those methods separately.

anarchivist commented on heidrun/nara_json.rb:
```diff
@@ -0,0 +1,575 @@
+  return contributors.first['contributor']['termName'] \
+    unless contributors.empty?
+
+  # FIXME:  It's not clear how to indicate the concept of "nothing" to
```
I'm not sure if this is the right approach, but I'm not sure what the best way to tackle it would be. A couple of options: 
- You could have something in the mapping reject any `nil` values. 
- You could add similar logic in `#make_contributor`, perhaps in the `contributors.select! ...` block. 
- The mapping DSL might need to have additional logic to understand what it means when it receives `nil` for a given property; e.g. if there are no other properties with URI or Literal values, nor does it have an associated URI, it is treated as an empty object. This might be a bigger conversation for when @no-reply returns from vacation.

no-reply replied:
My sense is that @anarchivist's first option is right. You should simply be filtering out values that are empty before this point. Here and elsewhere, there are a lot of empty nodes being created, leading to a bunch of data like this:

``` turtle
<http://purl.org/dc/terms/contributor> [ a <http://www.europeana.eu/schemas/edm/Agent>];
    <http://purl.org/dc/terms/isPartOf> [
      a <http://purl.org/dc/dcmitype/Collection>;
      <http://purl.org/dc/terms/title> "General Records of the Department of State"
    ];
    <http://purl.org/dc/terms/language> [ a <http://www.w3.org/2004/02/skos/core#Concept>];
    <http://purl.org/dc/terms/publisher> [ a <http://www.europeana.eu/schemas/edm/Agent>];
    <http://purl.org/dc/terms/spatial> [ a <http://www.europeana.eu/schemas/edm/Place>];
```

This says that there is a contributor, place, agent, etc..., each with no properties, that are affiliated with this object.

There's a way to do this, but I'm digging through these to try to understand what's going on.

no-reply replied:
The best example I can find for this is here:

https://github.com/dpla/heidrun-mappings/pull/16/files#diff-1e2b1d1a0a1dbcafbf93eb16e16bb19cR49

markbreedlove replied:
By putting the filtering code inside of these methods I was trying to keep the statements for each field shorter, so I understand how `select` or `reject` could probably be used lower down, but my concern then is that those statements would become more complex.  Also, the `contributor` field requires that if there are more than one Publisher, that the "Most Recent" one be used, so that seemed like a logic more complex than I'd want to put in one of those chained expressions.

no-reply replied:
> but my concern then is that those statements would become more complex.

That's interesting... my sense is that the goal with these mappings is to express as much as possible through the DSL methods.

We should probably talk about which of these prevailing principles we prefer as a group.

markbreedlove replied:
I definitely want to talk as a group about it -- especially with @guegueng and @AudreyAltman, who have not dealt yet with this mapping (and @guegueng not having as much of a Ruby background) because this is overwhelmingly a Ruby programming experience.  If I need detailed knowledge of how Ruby array methods work, how procs work, and so forth, then it seems I should follow various Ruby coding guidelines, like breaking things up into smaller methods.

The other thing with regard to the goals of the mappings is that I've had little documentation to go by, and I don't know where this goal is stated that it might guide my work.

anarchivist commented on heidrun/nara_json.rb:
```diff
@@ -0,0 +1,575 @@
+  # The 'type' property can be missing.
+  # See https://catalog.archives.gov/api/v1?naIds=65523&pretty=true&resultTypes=item,fileUnit&objects.object.@objectSortNum=1
+  type = (node.include? 'type') ? node['type']['termName'] : nil
+  [type, node['number']].compact.join ': '
```
I'm not sure if this is more idiomatic, but I might do something like `"#{type ? type + ": " : nil}#{node['number']}"`. I don't have a strong feeling about this, and it'd probably benefit from someone else's comments. :)

anarchivist commented on heidrun/nara_json.rb:
```diff
@@ -0,0 +1,575 @@
+#
+# FIXME:
+#
+# This may be wrong.  The original comment in this file was:
```
I think this is a question for @guegueng.

Gretchen: I'm not sure how NARA defines `coverageDates`, but my sense is that they are effectively like `<unitdate>` sets for EAD, which describe either inclusive or bulk dates for the grouping of archival material.

guegueng replied:
They are not temporal dates. As @anarchivist says, in EAD they are similar to bulk dates or inclusive dates...i.e., just a date range

anarchivist commented on heidrun/nara_json.rb:
```diff
@@ -0,0 +1,575 @@
+    #   <title>[VALUE]</title>
+    #   <recordGroupNumber>[VALUE]</recordGroupNumber>
+    # </parentRecordGroup> OR </parentCollection>
+    collection :class => DPLA::MAP::Collection do
```
@guegueng @markbreedlove Do we expect only 1 associated parent collection or record group?

anarchivist commented on heidrun/nara_json.rb:
```diff
@@ -0,0 +1,575 @@
+                         'parentRecordGroup | parentCollection', 'title')
+    end
+
+    contributor :class => DPLA::MAP::Agent do
```
Since `contributor` looks like it's likely to get an array, you probably want to use the `:each` argument at this level (e.g., see the [ESDN mapping lines 41-46](https://github.com/dpla/heidrun-mappings/blob/master/heidrun/esdn_mods.rb#L41-46).)

anarchivist commented on heidrun/nara_json.rb:
```diff
@@ -0,0 +1,575 @@
+    #     </creator>
+    #   </creatingOrganization> OR </creatingIndividual>
+    # </creatingOrganizationArray> OR </creatingIndividualArray>
+    creator :class => DPLA::MAP::Agent do
```
See [comment on `contributor`](https://github.com/dpla/heidrun-mappings/pull/14/files#r33056051).

markbreedlove replied:
I haven't figured out what to do with contributor yet, but this mapping for creator seems to work.

If I do this:

```
creator :class => DPLA::MAP::Agent,
        :each => record.field('description', 'item | itemAv | fileUnit',
                              'parentSeries',
                              'creatingOrganizationArray |' \
                              'creatingIndividualArray',
                              'creatingOrganization | creatingIndividual'),
        :as => :creator do
  providedLabel creator.field('creator', 'termName')
end
```

... I get this:

```
      <http://purl.org/dc/terms/creator> [
        a <http://www.europeana.eu/schemas/edm/Agent>;
        <http://dp.la/about/map/providedLabel> "Department of Defense. Department of the Navy. Bureau of Naval Personnel. 9/18/1947-1982",
          "Department of the Navy. Bureau of Navigation. 4/25/1913-5/1942"
      ],  [
        a <http://www.europeana.eu/schemas/edm/Agent>;
        <http://dp.la/about/map/providedLabel> "Department of Defense. Department of the Navy. Bureau of Naval Personnel. 9/18/1947-1982",
          "Department of the Navy. Bureau of Navigation. 4/25/1913-5/1942"
      ];
```

(Note the duplicate array)
... but if I use the code the way it is (no `:each`, just `record.field()`), I get:

```
      <http://purl.org/dc/terms/creator> [
        a <http://www.europeana.eu/schemas/edm/Agent>;
        <http://dp.la/about/map/providedLabel> "Department of Defense. Department of the Navy. Bureau of Naval Personnel. 9/18/1947-1982",
          "Department of the Navy. Bureau of Navigation. 4/25/1913-5/1942"
      ];
```

markbreedlove replied:
This is the relevant record:  https://catalog.archives.gov/api/v1?naIds=6004327&pretty=true&resultTypes=item,fileUnit&objects.object.@objectSortNum=1

no-reply replied:
Both results seem undesirable.  The first, for obvious reasons, the second because it says there is one creator, called both `"Department of Defense. Department of the Navy. Bureau of Naval Personnel. 9/18/1947-1982"` **and** `"Department of the Navy. Bureau of Navigation. 4/25/1913-5/1942"`.

The result we want to see is:

```
      <http://purl.org/dc/terms/creator> [
        a <http://www.europeana.eu/schemas/edm/Agent>;
        <http://dp.la/about/map/providedLabel> "Department of Defense. Department of the Navy. Bureau of Naval Personnel. 9/18/1947-1982" .
      ], 
      [
        a <http://www.europeana.eu/schemas/edm/Agent>;
        <http://dp.la/about/map/providedLabel> "Department of the Navy. Bureau of Navigation. 4/25/1913-5/1942" .
      ];
```

I think you should be able to get that with:

``` ruby
creator :class => DPLA::MAP::Agent,
        :each => record.field('description', 'item | itemAv | fileUnit',
                              'parentSeries',
                              'creatingOrganizationArray |' \
                              'creatingIndividualArray',
                              'creatingOrganization | creatingIndividual', 'creator'),
        :as => :creator do
  providedLabel creator.field('termName')
end
```

markbreedlove replied:
Hey, @no-reply, that code gets me this:

```
      <http://purl.org/dc/terms/creator> [
        a <http://www.europeana.eu/schemas/edm/Agent>;
        <http://dp.la/about/map/providedLabel> "Department of Defense. Department of the Navy. Bureau of Naval Personnel. 9/18/1947-1982",
          "Department of the Navy. Bureau of Navigation. 4/25/1913-5/1942"
      ],  [
        a <http://www.europeana.eu/schemas/edm/Agent>;
        <http://dp.la/about/map/providedLabel> "Department of Defense. Department of the Navy. Bureau of Naval Personnel. 9/18/1947-1982",
          "Department of the Navy. Bureau of Navigation. 4/25/1913-5/1942"
      ];
```

no-reply replied:
Well, that's not great.

Can you post the setup required to a Gist? I'd like to throw some prys in and figure out what's going on. I may be missing something, but that looks like it's not the result specified in the mapping.

markbreedlove replied:
I'll create a Gist.

I have also reverted Krikri to v0.5.7, which was before I made the mapping change to add the "or" operator, and I get the same result as pasted above, with repeated arrays.  (I modified the Gemfile and reinstalled the bundle.  Then I commented out everything in the mapping except for the `creator` section.  I modified the parameters to `field` to get rid of the `|` conditions, and ran the mapping manually on the same record that I'd used above.)

markbreedlove replied:
I should note that I did this on a local copy.  I did not revert anything in the repository!

markbreedlove replied:
Here is a Gist:  https://gist.github.com/markbreedlove/5c230307907246c32767
Is that helpful?  Is there anything I could add?

no-reply replied:
That's great. Can you add the opts you're using to instantiate the `NaraHarvester`?

markbreedlove replied:
It appears that I ran:

```
NaraHarvester.enqueue
```

The default options seem to work, if the NARA IDs file is in `/var/tmp/nara_ids`.

no-reply replied:
It looks to me like something is wrong with the `JsonParser` that's causing this. I'm trying to get to the bottom of it.

This is likely also the cause of the duplicated dates in a comment below.

anarchivist commented on heidrun/nara_json.rb:
```diff
@@ -0,0 +1,575 @@
+    #
+    # I expect `each`, below, to give me each of the proposableQualifiableDate
+    # objects one time only, but when when I examine the execution in `pry` it
+    # enters #make_data_provided_label nine times.  I expect it to enter that
```
I'm wondering if this is happening because of the `|` operator.

anarchivist replied:
How many nodes are returned by `:each` here, out of curiosity?

anarchivist commented on heidrun/nara_json.rb:
```diff
@@ -0,0 +1,575 @@
+      # self.end { |d| make_date_provided_label(d) }
+    end
+
+    temporal :class => DPLA::MAP::TimeSpan,
```
Does the same issue hold up for `temporal` as it does for `date`?

anarchivist commented on heidrun/nara_json.rb:
```diff
@@ -0,0 +1,575 @@
+    # </languageArray>
+    language :class => DPLA::MAP::Controlled::Language do
+
+      # FIXME:  shouldn't this be providedLabel?
```
Yes, I believe that it should be.

anarchivist commented on heidrun/nara_json.rb:
```diff
@@ -0,0 +1,575 @@
+    #  </geographicPlaceName>
+    # </geographicReferenceArray>
+    spatial :class => DPLA::MAP::Place do
+      providedLabel record.field('description', 'item | itemAv | fileUnit',
```
See [comment on `contributor`](https://github.com/dpla/heidrun-mappings/pull/14/files#r33056051). After looking at it more closely, maybe the logic should be like this:

```
spatial :class => DPLA::MAP::Place, 
  :each => record.field('description', 'item | itemAv | fileUnit', 'geographicReferenceArray'),
  :as => :place do
    providedLabel place.field('geographicPlaceName', 'termName')
end
```

markbreedlove replied:
See comment above, for `creator`; same results.

anarchivist commented on heidrun/nara_json.rb:
```diff
@@ -0,0 +1,575 @@
+                                 'geographicPlaceName', 'termName')
+    end
+
+    publisher :class => DPLA::MAP::Agent do
```
See [comment on `contributor`](https://github.com/dpla/heidrun-mappings/pull/14/files#r33056051) and [comment on `spatial`](https://github.com/dpla/heidrun-mappings/pull/14/files#r33057678).

anarchivist commented on heidrun/nara_json.rb:
```diff
@@ -0,0 +1,575 @@
+    # <termName>[VALUE]</termName>
+    # </topicalSubject>
+    # </topicalSubjectArray>
+    subject :class => DPLA::MAP::Concept do
```
See [comment on `contributor`](https://github.com/dpla/heidrun-mappings/pull/14/files#r33056051) and [comment on `spatial`](https://github.com/dpla/heidrun-mappings/pull/14/files#r33057678).

markbreedlove replied:
See comment for `creator` above; same results.

Commit:
Assign language providedLabel instead of prefLabel

no-reply commented on heidrun/nara_json.rb:
```diff
@@ -0,0 +1,573 @@
+  object :class => DPLA::MAP::WebResource do
+    uri record.field('objects', 'object', 'file').first_value
+              .map { |o| make_obj_uri(o) }
+    dcformat record.field('objects', 'object', 'file').first_value
```
I'm confused by a lot of this. Is there something that this is doing beyond:

``` ruby
dcformat record.field('objects', 'object', 'file').first_value.field('@mime')
```

?

markbreedlove replied:
I don't think so.  If it works that way, it should be changed.

no-reply replied:
:+1:

I have a branch that handles these as:

``` ruby
object :class => DPLA::MAP::WebResource do                                                                             
  uri record.field('objects', 'object', 'file')                                                                        
       .first_value.field('@url').map { |url| URI::escape(url.value) }                                                 
  dcformat record.field('objects', 'object', 'file')
                .first_value.field('@mime')                                                                                
end                                                                                                                    

preview :class => DPLA::MAP::WebResource do                                                                            
  uri record.field('objects', 'object', 'thumbnail')                                                                   
          .first_value.field('@url').map { |url| URI::escape(url.value) }                                                 
  dcformat record.field('objects', 'object', 'thumbnail')                                                              
              .first_value.field('@mime')                                                                                
end
```

Commit:
nara mapping updates

no-reply commented:
I wasn't able to identify any problems with the JSON parser, but I do think the problem we were attributing to it is fixed with some changes to the mapping.

I've added some inline comments to this commit, explaining what I'm up to: https://github.com/dpla/heidrun-mappings/commit/5768026c078ab1a037d7eb043f4abd9b72a9f289#diff-2dc9e1a949d2a5e9d0999c268b86dfcfR209

no-reply commented:
Assigned to @markbreedlove, looking for feedback on https://github.com/dpla/heidrun-mappings/pull/14#commitcomment-13144153

markbreedlove commented:
After analyzing this for a while, I have to say I feel ambivalent about how `if` or `if_field` would work.  Especially considering our schedule, I don't think the subtleties matter that much.

In the long run, we should be asking people outside our group who need to be enabled to do mappings what they think.  My overall reaction to the DSL and the refactoring that's being done here is that it matters more if you're concerned with doing things in a Ruby-orthodox way, because it still has Ruby's symbols and rules. Will the proposed functions be capable of generating error messages that point precisely to the cause of a given problem, the way the existing functions could?

Commit:
Fixup NARA mapping

 - Extension methods for NARA are lambda-ized and updated to use current
DSL features.
 - `&make_relation` is updated to address a more complex state of affairs
than is originally communicated in the spreadsheet; the
relation/collection hierarchy is more malleable than is described. This
is documented in the comments, but needs to be updated in the
spreadsheet mapping.
 - `&make_rights` is refactored.
 - The `collection` mapping is updated to address the collection hierarchy
as described above.
 - The `date` mapping is updated to handle `coverageDates` with begin and
end date nodes.

 - `description` is updated to correctly match `scopeAndContentNote` as
described in the mapping (values appear directly, rather than nested in
`generalNote.note`
 - `#field('termName')` calls are moved into their respective
`ChildDeclaration` blocks. This avoids splitting values in the case that
more than one `termName` is given. As far as I know, this doesn't
actually happen in production data, but it's slightly safer with the
adjustment.
 - Removes XML formatted comments. These comments are duplicated from the
mapping spreadsheet; including them here spreads maintenance. They are
also somewhat misleading, since the values retrieved are JSON.

no-reply commented:
Replaced by #49. Any further edits should happen there.
